File: Data Structures & Algorithms/max-water-container/submission-1.java
class Solution {
    public int maxArea(int[] heights) {
        int i=0;
        int j = heights.length -1;

        int maxContainer =-1;

        while(i < j) {
            int iH = heights[i];
            int jH = heights[j];
            int currContainer = (j-i) * Math.min(iH, jH);

            //System.out.printf("i: %d, j: %d, iH: %d, jH: %d\n", i, j, iH, jH);

            if(currContainer > maxContainer) {
                maxContainer = currContainer;

               // System.out.printf("Max container: %d  i:%d j:%d  %dx%d \n", maxContainer, i, j, iH, jH);
            }

            if(iH < jH) {
                i++;
            } else {
                j--;
            }
        }

        return maxContainer;
    }
}

File: Data Structures & Algorithms/largest-rectangle-in-histogram/submission-1.kt
class Solution {
       fun largestRectangleArea(heights: IntArray): Int {
        /*
        maxArea = 0
        stack = pusty
        heights' = heights + [0]
        for i in 0 .. heights'.length-1:
            while stack not empty AND heights'[stack.top] > heights'[i]:
                top = stack.pop()
                w = stack empty ? i : i - stack.top - 1
                maxArea = max(maxArea, heights'[top] * w)
            stack.push(i)
        return maxArea
*/
        val stack = ArrayDeque<Int>()
        var maxArea:Int = 0
        val heightsPrim= heights.copyOf().plus(0)

        for(i in 0..heightsPrim.size-1 ) {
            val last = if (stack.isNotEmpty()) stack.last() else null
            println("i: $i, stack.last():${last} height: ${heightsPrim[i]}")
            while (stack.isNotEmpty() && heightsPrim[stack.last()] > heightsPrim[i]) {
                val top = stack.removeLast()
                val w = if(stack.isEmpty())  i else i - stack.last()-1

                println("top: $top, maxArea: ${maxOf(maxArea, heightsPrim[top] * w)}")
                maxArea = maxOf(maxArea, heightsPrim[top] * w)
            }
            stack.add(i)
            println("--------------")
        }

        return maxArea
    }
}
